File: Medium/200.number-of-islands/solution2.rb
# timestamp: 2023-01-04T09:19:02+08:00

# question: [Number of Islands](https:/leetcode.com/problems/number-of-islands)
# train of thought
TRAIN_OF_THOUGHT = <<DOC.freeze
  改成 DFS 就過了
  Run time 150 ms Beats 100%
  Memory 219.5 MB Beats 69.23%
  酷
DOC

# @param {Character[][]} grid
# @return {Integer}
def num_islands(grid)
  count = 0
  # puts "======================"
  0.upto(grid.size-1) do |i|
    0.upto(grid[0].size-1) do |j|
      if grid[i][j] == "1"
        find_connected(grid, i, j)
        count += 1
      end
    end
  end
  count
end

def find_connected(grid, i, j)
  grid[i][j] = "2"
  if j < grid[i].size - 1 && grid[i][j+1]=="1"
    find_connected(grid, i, j+1)
  end
  # >
  if i < grid.size - 1 && grid[i+1][j]=="1"
    find_connected(grid, i+1, j)
  end
  # V
  if i < grid.size - 1 && grid[i+1][j]=="1"
    find_connected(grid, i+1, j)
  end
  # <
  if j > 0 && grid[i][j-1]=="1"
    find_connected(grid, i, j-1)
  end
  # ^
  if i > 0 && grid[i-1][j]=="1"
    find_connected(grid, i-1, j)
  end
end

File: Medium/200.number-of-islands/solution2.ts
// 改成 DFS 就行了，也試試看 Typescript
function numIslands(grid: string[][]): number {
  let count = 0;
  for(let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++){
      if (grid[i][j] === '1') {
        findConnected(grid, i, j);
        count += 1
      }
    }
  }
  return count;
};

function findConnected(grid: string[][], i: number, j: number): void{
  grid[i][j] = '2';
  if (j < grid[0].length - 1 && grid[i][j+1] === '1') {
    findConnected(grid, i,j+1);
  }
  if (i < grid.length - 1 && grid[i+1][j] === '1') {
    findConnected(grid, i+1,j);
  }
  if (j > 0 && grid[i][j-1] === '1') {
    findConnected(grid, i,j-1)
  }
  if (i > 0 && grid[i-1][j] === '1') {
    findConnected(grid, i-1,j)
  }
}